### ISSUE TITLE
Children props are lost when custom element is moved in the DOM

### ISSUE THREAD

**Issue by fab920** [jahilldev/component-elements#18]
Components are hydrated each time they are appended to the document.

```
var c = document.createElement("my-test");
document.body.appendChild(c);
document.body.appendChild(c);
```
The render() function is called twice. The second time e.g. this.props.children does not work any more.

**Comment by jahilldev** [comment]
Hi @fab920!

Thanks for reaching out.

If I understand correctly, you're pointing out that each custom element instance added to the DOM is an independently rendered component? If this is the case, this is intended behaviour. Each custom element should have its own separate context, state, etc, and will require it's own render.

If I'm wrong, and this isn't what you're pointing out, would you mind clarifying for me?

It would also be really useful if you could you provide me with a simple reproduction, a quick repo that demonstrates the issue would be really helpful 👍 

Thanks!

**Comment by fab920** [comment]
Hi @jahilldev ,
no, I want to point out that an instance of an custom element is hydrated each time it is added to the DOM.

I would provide a JsFidle if I knew how to create one with preactement, but if you just execute the 3 lines of codein the initial post, e.g. in the developer console, with any custom component, the problem should be be reproducible.

**Comment by jahilldev** [comment]
> I want to point out that an instance of an custom element is hydrated each time it is added to the DOM.

So this seems like what I would expect. For example, if I'm dynamically creating an element, `my-test` let's say, whenever I add it to the DOM via `appendChild`, that is essentially a new component instance being instantiated, which will require its own render / hydration.

When you make use of a component multiple times in (p)react, you wouldn't expect all of those components to share the same context, state, etc. They are essentially separate entities.

If I run your snippet on a site that makes use of `(p)reactment`, I can see it adding the custom element to the DOM, and rendering the component. I don't see a double render however, as an element can only be added to the DOM once:

![Screenshot 2022-04-26 at 12 45 39](https://user-images.githubusercontent.com/13650046/165293169-93d259c2-f945-4980-bb64-f1aa8871da19.png)

I feel like I'm probably missing something that an example reproduction would make obvious. I'd suggest creating a repo and outlining exactly what you're expecting to happen, and the issue you're highlighting. 

Thanks!

**Comment by fab920** [comment]
> So this seems like what I would expect. For example, if I'm dynamically creating an element, <my-test /> let's say, whenever I > add it to the DOM via appendChild, that is essentially a new component instance being instantiated, which will require its own > render / hydration.

No, that's not what should be happening. When you create a DOM node (normal or a component) and add it to the DOM multiple times to the node gets moved around in the DOM.

_The appendChild() method of the [Node](https://developer.mozilla.org/en-US/docs/Web/API/Node) interface adds a node to the end of the list of children of a specified parent node. If the given child is a reference to an existing node in the document, appendChild() moves it from its current position to the new position (there is no requirement to remove the node from its parent node before appending it to some other node)._
Source: https://developer.mozilla.org/en-US/docs/Web/API/Node/appendChild

I probably should have chosen a better wording and a better example :-(

```
var c = document.createElement("my-test"); 
document.getElementById("NodeA").appendChild(c);
document.getElementById("NodeA").removeChild(c);
document.getElementById("NodeB").appendChild(c);
```

So essentially the component is recreated every time you move it around in the DOM.


Gist with a more complete example 
https://gist.github.com/fab920/7340b4a19ed11773e6f4647e889fedd0

**Comment by fab920** [comment]
The same happens when you move one of its ancestor elements

```
var c = document.createElement("my-test"); 
document.getElementById("NodeA").appendChild(c);

document.getElementById("NodeB").appendChild(document.getElementById("NodeA"));
```
will recreate (and rerender) my-test.

**Comment by jahilldev** [comment]
@fab920 Ok, well that makes a lot more sense! 

Your initial example is as I said though, you can only add it to the DOM once in that position. The example you've provided now is completely different!

Give me a few days to check out your example and I'll see what's going on.

**Comment by jahilldev** [comment]
@fab920 A quick question before we go on, I'm interested to know the use case here. What are you trying to achieve?

I can see why this is happening, when you move the Custom Element, the `onDisconnected` callback is being invoked, which "unmounts" the component, and the "remounts" when it's added back in at another position, firing `onConnected`. 

Why is the re-render an issue when it's being moved?

**Comment by fab920** [comment]
In my case it's about the content in the `<slots/>`. 

When the component is recreated `this.props.children` is empty. If you look at my gist, the `lorum ipsum` is gone after the component is recreated.

**Comment by jahilldev** [comment]
@fab920 So I've created a simple repro for you below:
https://github.com/jahilldev/preactement-moved

I still don't fully understand what you're referring to about slots, these are correctly persisted when the element is moved in the DOM. Could you pull and run my repro and feedback, please?

`this.props.children` _are_ lost however, for the reason I outline above. I'm going to open a PR shortly to address this.

I'd _strongly_ recommend going to the effort of creating a runnable repro of an issue you're raising in future, though. It's really helpful to isolate and describe the problem, and 9 times out of 10, the maintainer won't do this for you.

I'll let you know once the PR is up.

Thanks!

**Comment by fab920** [comment]
I'm not moving the component itself, but it just gets moved as part of one of its ancestors moving. 
A lot of third party libraries & components tend to move dom nodes to the `body'' for modals, popups, popovers, ...

E.g. we are using preact together with ionic. 
```
render(
	<ion-modal is-open="true">
		<my-test>lorum ipsum</my-test>
	</ion-modal>, document.body
);
```
is not working. `lorum ipsum` is missing from the output. 

And, on other problem is, that if the code is generated server side, I can not just recreate the `my-test`. The needed data is missing on the client.

**Comment by jahilldev** [comment]
@fab920 Could you verify you've looked at my above comment?

**Comment by fab920** [comment]
Looking right now. Posted my answer just the same moment you did.

Thanks for all your help.

**Comment by fab920** [comment]
>  still don't fully understand what you're referring to about slots, these are correctly persisted when the element is moved in the DOM. Could you pull and run my repro and feedback, please?

Sorry!! - Indeed '''<slots/>''' are working as expected and it's only the children, that are not working.

> I'd strongly recommend going to the effort of creating a runnable repro of an issue you're raising in future, though. It's really helpful to isolate and describe the problem, and 9 times out of 10, the maintainer won't do this for you.
Sorry, I tried to and could provide one now :-/ . It just took me ages to set it up. Starting a new project from scratch is something I never had to do until now.

Thanks again for your effort and the fast PR

**Comment by jahilldev** [comment]
@fab920 Ok great, we got there in the end, go team 🥳

No worries, I'm just glad we've got a solution now.

Will let you know once the PR is up 👍

**Comment by jahilldev** [comment]
@fab920 Here we go:
https://github.com/jahilldev/component-elements/pull/19

**Comment by jahilldev** [comment]
@fab920 I've just published `preactement v1.8.2` and `reactement v1.1.2`, you can go ahead and update.

Let me know how it goes, please feedback either way!

Thanks 👍

**Comment by fab920** [comment]
Looking good.

Thanks a lot!
